### ISSUE TITLE
C# Interfaces can't be used as arguments in signals

### ISSUE THREAD

**Issue by felaugmar** [godotengine/godot#38188]
**Godot version:** 3.2.1 stable mono

**OS/device including version:** Windows 10

**Issue description:**
C# classes needs to be derived from `Godot.Object` to be used, but what about Interfaces?

**Steps to reproduce:**
Assume this C# Interface:

``` C#
public interface ITransitionable
{
  void MoveTo(string mapPath, string checkpointName);
}
```

And connect a signal to a C# method expecting it as the first argument:

```C#
public void OnEntered(ITransitionable transitionable)
{
  transitionable.MoveTo(MapPath, CheckpointName);
}
```

This will not work and an error raises: **variant_to_mono_object: Attempted to convert Variant to an unmarshallable managed type. Name: 'ITransitionable' Encoding: 18.**

But doing like below will work just fine:

```C#
public void OnEntered(Godot.Object @object)
{
  var transitionable = @object as ITransitionable;

  transitionable.MoveTo(MapPath, CheckpointName);
}
```

**Comment by Mek101** [comment]
I think this happens because in C# interfaces don't exist as runtime objects. Any standard .NET vm only supports value types like structs or reference types like classes, as interfaces exist as a compiler construct only

**Comment by Calinou** [comment]
cc @neikeq

**Comment by neikeq** [comment]
This happens because we currently only support Godot.Object and other types convertible to Variant.

**Comment by felaugmar** [comment]
The actual type derives from `Godot.Object`, asking for an interface is just a contract the type should met.

**Comment by raymoo** [comment]
You could make `OnEntered` a generic function with a constraint on the parameter that it has to satisfy the interface. For example:
```
public void OnEntered<T>(T transitionable) where T : Godot.Object, ITransitionable
{
  transitionable.MoveTo(MapPath, CheckpointName);
}
```

**Comment by jjmontesl** [comment]
I just wanted to reflect here that this also impacts ability to await for async functions called via `Godot.Object.Call`, ask `Task`cannot be marshalled:

```
Attempted to convert an unmarshallable managed type to Variant. Name: 'Task`1'
```

(This is detailed in the referenced issue #41358)

**Comment by felaugmar** [comment]
Can anyone confirm if this is still an issue in **Godot 4.x**?

**Comment by OpsisKalopsis** [comment]
> Can anyone confirm if this is still an issue in **Godot 4.x**?

I haven't tested it but reading from the documentation it doesn't  seems possible.

https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_signals.html

_"Signals support arguments of any [Variant-compatible](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_variant.html#doc-c-sharp-variant) type.

Consequently, any Node or Reference will be compatible automatically, but custom data objects will need to inherit from GodotObject or one of its subclasses."_

**Comment by dreamsComeTrue** [comment]
Yep, confirming this in January of 2024 - current version: 4.3 dev2 STILL doesn't support interfaces as Signal's arguments (which is a bummer since it's highly important).

**Comment by BananaHolograma** [comment]
I had the same problem today and in the Discord csharp channel helped me with this workaround. I needed to pass an interface that I named `**IInteractor**` so the trick here is type the handler argument as `GodotObject` and check your desired type in the Listener:

    [Signal]
    public delegate void InteractedEventHandler(GodotObject interactor);
    
      private void OnInteracted(GodotObject interactorParameter)
    {
        if (interactorParameter is IInteractor interact)
        {
           //....
        }
     }
